lib/page/set-uploads/set-uploads.js

```diff
@@ -1,43 +1,133 @@
+        if (!hintValue) {
+          // Strip string before slash - eg. image/svg+xml => svg.xml
+          typeLookup = type.replace(/.*\//, '').replace(/\+.*/g, '')
+          if (typeLookup === '*') {
+            // Wildcard - revert to string before slash - eg. image/* => image
+            typeLookup = type.replace(/\/.*/, '')
+          }
+          hintValue = getString(`fileupload.hint.type.${typeLookup}`, userData.contentLang) || typeLookup
+        }
+        arr[index] = hintValue
+      })
+      control.acceptHints = acceptHints
+    }
+
     if (control.accept) {
       control.accept = control.accept.join(',')
     }
     const maxFiles = getUploadMaxFiles(control)
+    const minFiles = getUploadMinFiles(control)
     const currentFiles = getUploadFiles(control, userData)
     const currentFileCount = getUploadFileCount(control, userData)
     const maxAvailableSlots = maxFiles - currentFileCount
     let slots = 0
     if (maxAvailableSlots) {
-      let uploadSlots = req.body[`${control.name}_slots`] || 1
-      const addFileInvoked = req.body.addFile === control.name
+      const body = userData.getBodyInput()
+      const {addFile, removeSlot} = body
+      let uploadSlots = body[`${control.name}_slots`] || 1
+      const addFileInvoked = addFile === control.name
       if (addFileInvoked) {
         uploadSlots++
       }
+      if (removeSlot) {
+        const removeSlotInvoked = getNormalisedUploadControlName(removeSlot) === control.name
+        if (removeSlotInvoked) {
+          uploadSlots--
+        }
+      }
+      uploadSlots = Math.max(uploadSlots, minFiles)
       const successfulUploads = userData.getSuccessfulUploadsCount(control.name)
       uploadSlots = uploadSlots - successfulUploads
       slots = Math.min(uploadSlots, maxAvailableSlots)
       control.slots = slots
       if (slots < maxAvailableSlots) {
-        control.addFile = true
+        control.addFile = getString('fileupload.file.button.add', userData.contentLang)
       }
     }
 
-    const fileRemoveText = getString('fileupload.file.remove', pageInstance.contentLang)
-    const slotLabel = getString('fileupload.slot.label', pageInstance.contentLang)
+    if (!control.hint) {
+      let controlHintString = getString('fileupload.hint', userData.contentLang)
+      controlHintString = format(controlHintString, {
+        maxfiles: control.maxFiles || 1,
+        maxsize: control.maxSize,
+        types: control.acceptHints
+      }, {
+        markdown: false,
+        lang: userData.contentLang
+      })
+      control.hint = controlHintString
+    }
+
+    if (slots) {
+      let uploadExplanation = getString('fileupload.hint.timing', userData.contentLang)
+      if (uploadExplanation) {
+        uploadExplanation = format(uploadExplanation, {
+          count: slots
+        }, {
+          markdown: false,
+          lang: userData.contentLang
+        })
+        control.hint += '\n\n'
```

Comment:
Any workaround with this?

Reply:
Yes - we can look in to doing this totally in the data